### PR TITLE
Add masking to BatchNormalization

### PR DESCRIPTION
**Pull request by TomCoded**
Proposed solution for the bug reported in https://github.com/keras-team/keras/issues/18759

Prior to this, masking was not effective in batch normalization layers. This approach pulls weighted_moments and divide_no_nan logic into batch_normalization using core operations so that masking will work with any back-end instead of just tensorflow.

Open to suggestions.

### CODE REVIEW

File: keras/layers/normalization/batch_normalization.py [hunk 1]
```diff
@@ -282,3 +287,57 @@ def get_config(self):
...
+        weighted_distsq = ops.sum(
+            mask_weights_broadcasted * squared_difference,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        variance = self._divide_no_nan(weighted_distsq, sum_of_weights)
+
+        return ops.squeeze(mean), ops.squeeze(variance)
+
+    def _divide_no_nan(self, x1, x2):
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:332]: I am not convinced that this is the best solution performance wise (it could be fine, not sure). Would it be better to just divide in float and use an epsilon fuzz factor to avoid division by zero?

File: keras/layers/normalization/batch_normalization.py [hunk 2]
```diff
@@ -282,3 +287,57 @@ def get_config(self):
...
+        original_dtype = x1.dtype
+        x2_binary = ops.cast(x2, dtype="bool")
+        x2_mask = ops.cast(x2_binary, dtype=original_dtype)
+        x1_masked = x1 * x2_mask
+
+        x2_binary_inverted = ~x2_binary
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:338]: Prefer using an op rather than a unary operator since it makes it easy to parse what code it is executing

File: keras/layers/normalization/batch_normalization.py [hunk 3]
```diff
@@ -84,6 +84,12 @@ class BatchNormalization(Layer):
             - `training=False`: The layer will normalize its inputs using
             the mean and variance of its moving statistics, learned during
             training.
+        mask: Binary tensor of shape broadcastable to `inputs` tensor, with
+            True values indicating the positions for which mean and variance
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:88]: Use backticks for `True`

**Review by fchollet — COMMENTED** [review]
Thank you for the contribution!

File: keras/layers/normalization/batch_normalization.py (hunk 3 above)
Comment (TomCoded) [reply keras/layers/normalization/batch_normalization.py]: Done.

File: keras/layers/normalization/batch_normalization.py (hunk 1 above)
Comment (TomCoded) [reply keras/layers/normalization/batch_normalization.py]: Yes! Removed in favor of an epsilon approach, good call. The way this is used here mitigates the risks of that approach from negative epsilon values in the denominator (which is generated from the binary masks) and misleadingly large results coming from division by small epsilon (because the numerators are masked).

File: keras/layers/normalization/batch_normalization.py (hunk 2 above)
Comment (TomCoded) [reply keras/layers/normalization/batch_normalization.py]: Removed.

File: keras/layers/normalization/batch_normalization.py [hunk 4]
```diff
@@ -282,3 +287,44 @@ def get_config(self):
...
+            weighted_inputs,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        sum_of_weights = ops.sum(
+            mask_weights_broadcasted,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        mean = weighted_input_sum / (sum_of_weights + self.epsilon)
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:319]: An issue here is that `self.epsilon` is the batch norm epsilon, used to compute the inverse of the variance in the `batch_normalization` op. It is usually set to a fairly large value -- for instance, the default value is 1e-3, and some folks use 1e-2. So the moment computation here will be very imprecise.

Prefer using `config.epsilon()`, which is typically 1e-7.

Comment (TomCoded) [reply keras/layers/normalization/batch_normalization.py:319]: That makes sense. Fixing, thanks!

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!
